File: Masjidly - Official Masjid Prayer Times/Domain/PrayerModels.swift
import Foundation

struct DataRevision: Codable, Equatable, Sendable {
    let dataRevision: Double
    let updatedAt: Double
}

struct PrayerDataVersions: Codable, Equatable, Sendable {
    let mosquesUpdatedAt: Double
    let monthlyUpdatedAt: Double
    let ramadanUpdatedAt: Double
    let dstUpdatedAt: Double
}

struct CachedPrayerDataVersions: Codable, Equatable, Sendable {
    let versions: PrayerDataVersions
    let checkedAt: Date
}

struct Mosque: Codable, Equatable, Identifiable, Sendable {
    let id: String
    let name: String
    let address: String
    let lat: Double
    let lng: Double
    let slug: String
    let citySlug: String?
    let cityName: String?
    let countryCode: String?
    let countryName: String?
    let timezone: String?
    let website: String?
    let isHidden: Bool?

    var isHiddenResolved: Bool { isHidden ?? false }
    var cityDisplayName: String { cityName ?? "Sheffield" }

    /// Groups mosques for city pickers; stable for `SettingsStore.selectedCityGroupingKey`.
    var cityGroupingKey: String {
        if let s = citySlug, !s.isEmpty { return "slug:\(s)" }
        let label = cityName ?? cityDisplayName
        return "name:\(label.lowercased())"
    }
}

struct PrayerTime: Codable, Equatable, Sendable {
    let date: Int
    let fajr: String
    let shurooq: String
    let dhuhr: String
    let asr: String
    let asrMithl2: String?
    let maghrib: String
    let isha: String

    enum CodingKeys: String, CodingKey {
        case date, fajr, shurooq, dhuhr, asr, maghrib, isha
        case asrMithl2 = "asr_mithl2"
    }

    init(date: Int, fajr: String, shurooq: String, dhuhr: String, asr: String, asrMithl2: String? = nil, maghrib: String, isha: String) {
        self.date = date
        self.fajr = fajr
        self.shurooq = shurooq
        self.dhuhr = dhuhr
        self.asr = asr
        self.asrMithl2 = asrMithl2
        self.maghrib = maghrib
        self.isha = isha
    }
}

struct IqamahTimeRange: Codable, Equatable, Sendable {
    let dateRange: String
    let fajr: String
    let dhuhr: String
    let asr: String
    let maghrib: String?
    let isha: String
    let jummah: String?

    enum CodingKeys: String, CodingKey {
        case dateRange = "date_range"
        case fajr, dhuhr, asr, maghrib, isha, jummah
    }

    init(dateRange: String, fajr: String, dhuhr: String, asr: String, maghrib: String?, isha: String, jummah: String?) {
        self.dateRange = dateRange
        self.fajr = fajr
        self.dhuhr = dhuhr
        self.asr = asr
        self.maghrib = maghrib
        self.isha = isha
        self.jummah = jummah
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        dateRange = try c.decode(String.self, forKey: .dateRange)
        fajr = try Self.decodeIqamahValue(c, forKey: .fajr)
        dhuhr = try Self.decodeIqamahValue(c, forKey: .dhuhr)
        asr = try Self.decodeIqamahValue(c, forKey: .asr)
        maghrib = try Self.decodeOptionalIqamahValue(c, forKey: .maghrib)
        isha = try Self.decodeIqamahValue(c, forKey: .isha)
        jummah = try Self.decodeOptionalIqamahValue(c, forKey: .jummah)
    }

    private static func decodeIqamahValue(_ c: KeyedDecodingContainer<CodingKeys>, forKey key: CodingKeys) throws -> String {
        if let value = try? c.decode(String.self, forKey: key) { return value }
        if let values = try? c.decode([String].self, forKey: key) { return values.joined(separator: ", ") }
        throw DecodingError.typeMismatch(String.self, DecodingError.Context(codingPath: c.codingPath + [key], debugDescription: "Expected string or string array"))
    }

    private static func decodeOptionalIqamahValue(_ c: KeyedDecodingContainer<CodingKeys>, forKey key: CodingKeys) throws -> String? {
        guard c.contains(key), !(try c.decodeNil(forKey: key)) else { return nil }
        if let value = try? c.decode(String.self, forKey: key) { return value }
        if let values = try? c.decode([String].self, forKey: key) { return values.joined(separator: ", ") }
        return nil
    }
}

struct MonthPrayerData: Codable, Equatable, Sendable {
    let month: String
    let prayerTimes: [PrayerTime]
    let iqamahTimes: [IqamahTimeRange]
    let jummahIqamah: String

    enum CodingKeys: String, CodingKey {
        case month
        case prayerTimes = "prayer_times"
        case iqamahTimes = "iqamah_times"
        case jummahIqamah = "jummah_iqamah"
    }
}

struct RamadanPrayerDay: Codable, Equatable, Sendable {
    let ramadanDay: Int
    let gregorian: String
    let fajr: String
    let shurooq: String
    let dhuhr: String
    let asr: String
    let asrMithl2: String?
    let maghrib: String
    let isha: String

    enum CodingKeys: String, CodingKey {
        case ramadanDay = "ramadan_day"
        case gregorian, fajr, shurooq, dhuhr, asr, maghrib, isha
        case asrMithl2 = "asr_mithl2"
    }

    init(ramadanDay: Int, gregorian: String, fajr: String, shurooq: String, dhuhr: String, asr: String, asrMithl2: String? = nil, maghrib: String, isha: String) {
        self.ramadanDay = ramadanDay
        self.gregorian = gregorian
        self.fajr = fajr
        self.shurooq = shurooq
        self.dhuhr = dhuhr
        self.asr = asr
        self.asrMithl2 = asrMithl2
        self.maghrib = maghrib
        self.isha = isha
    }
}

struct RamadanPrayerData: Codable, Equatable, Sendable {
    let month: String
    let gregorianStart: String
    let gregorianEnd: String
    let prayerTimes: [RamadanPrayerDay]
    let iqamahTimes: [IqamahTimeRange]
    let jummahIqamah: String

    enum CodingKeys: String, CodingKey {
        case month
        case gregorianStart = "gregorian_start"
        case gregorianEnd = "gregorian_end"
        case prayerTimes = "prayer_times"
        case iqamahTimes = "iqamah_times"
        case jummahIqamah = "jummah_iqamah"
    }
}

struct DailyPrayerTimes: Codable, Equatable, Sendable {
    var date: String
    var fajr: String
    var sunrise: String
    var dhuhr: String
    var asr: String
    var maghrib: String
    var isha: String
}

struct DailyIqamahTimes: Codable, Equatable, Sendable {
    var fajr: String
    var dhuhr: String
    var asr: String
    var maghrib: String
    var isha: String
    var jummah: String
}

struct UkDstYear: Codable, Equatable, Sendable {
    let year: Int
    let startDate: String
    let endDate: String

    enum CodingKeys: String, CodingKey {
        case year
        case startDate = "start_date"
        case endDate = "end_date"
    }
}

struct UkDstCalendar: Codable, Equatable, Sendable {
    let ukDstDates: [UkDstYear]

    enum CodingKeys: String, CodingKey {
        case ukDstDates = "uk_dst_dates"
    }
}

File: apps/android/app/src/main/kotlin/com/mikhailspeaks/masjidly/domain/PrayerModels.kt
package com.mikhailspeaks.masjidly.domain

import java.time.Duration
import java.time.Instant
import kotlin.math.floor

/** Default mosque slug (matches iOS `MosqueDefaults.defaultSlug`). */
object MosqueDefaults {
    const val DEFAULT_MOSQUE_SLUG = "muslim-welfare-house"
}

data class Mosque(
    val id: String,
    val name: String,
    val address: String,
    val lat: Double,
    val lng: Double,
    val slug: String,
    val citySlug: String? = null,
    val cityName: String? = null,
    val countryCode: String? = null,
    val countryName: String? = null,
    val timezone: String? = null,
    val website: String? = null,
    val isHidden: Boolean? = null,
) {
    val isHiddenResolved: Boolean get() = isHidden ?: false
    val cityDisplayName: String get() = cityName ?: "Sheffield"

    /** Groups mosques for city pickers; stable for settings `selectedCityGroupingKey`. */
    val cityGroupingKey: String
        get() {
            val s = citySlug
            if (!s.isNullOrEmpty()) return "slug:$s"
            val label = cityName ?: cityDisplayName
            return "name:${label.lowercase()}"
        }
}

data class PrayerTime(
    val date: Int,
    val fajr: String,
    val shurooq: String,
    val dhuhr: String,
    val asr: String,
    val asrMithl2: String? = null,
    val maghrib: String,
    val isha: String,
)

data class IqamahTimeRange(
    val dateRange: String,
    val fajr: String,
    val dhuhr: String,
    val asr: String,
    val maghrib: String? = null,
    val isha: String,
    val jummah: String? = null,
)

data class MonthPrayerData(
    val month: String,
    val prayerTimes: List<PrayerTime>,
    val iqamahTimes: List<IqamahTimeRange>,
    val jummahIqamah: String,
)

data class RamadanPrayerDay(
    val ramadanDay: Int,
    val gregorian: String,
    val fajr: String,
    val shurooq: String,
    val dhuhr: String,
    val asr: String,
    val asrMithl2: String? = null,
    val maghrib: String,
    val isha: String,
)

data class RamadanPrayerData(
    val month: String,
    val gregorianStart: String,
    val gregorianEnd: String,
    val prayerTimes: List<RamadanPrayerDay>,
    val iqamahTimes: List<IqamahTimeRange>,
    val jummahIqamah: String,
)

data class DailyPrayerTimes(
    var date: String,
    var fajr: String,
    var sunrise: String,
    var dhuhr: String,
    var asr: String,
    var maghrib: String,
    var isha: String,
)

data class DailyIqamahTimes(
    var fajr: String,
    var dhuhr: String,
    var asr: String,
    var maghrib: String,
    var isha: String,
    var jummah: String,
)

data class UkDstYear(
    val year: Int,
    val startDate: String,
    val endDate: String,
)

data class UkDstCalendar(
    val ukDstDates: List<UkDstYear>,
)

enum class AsrIqamahPreference {
    FIRST,
    SECOND,
    ;

    companion object {
        fun fromWire(value: String?): AsrIqamahPreference =
            when (value?.lowercase()) {
                "second" -> SECOND
                else -> FIRST
            }
    }

    val wireValue: String
        get() = when (this) {
            FIRST -> "first"
            SECOND -> "second"
        }
}

typealias AsrTimingPreference = AsrIqamahPreference

data class NextPrayerCountdownResult(
    val nextName: String,
    val nextTime: String,
    val totalSeconds: Int,
    val isIqamah: Boolean,
    val isJummah: Boolean,
) {
    val hours: Int get() = totalSeconds / 3600
    val minutes: Int get() = (totalSeconds % 3600) / 60
    val seconds: Int get() = totalSeconds % 60
}

enum class HeroCountdownLabelKind {
    ADHAN_IN,
    IQAMAH_IN,
    NEXT_PRAYER,
    ;

    val wireValue: String
        get() = when (this) {
            ADHAN_IN -> "adhanIn"
            IQAMAH_IN -> "iqamahIn"
            NEXT_PRAYER -> "nextPrayer"
        }
}

/** Widget-aligned interval for the home hero progress ring (`MasjidlyWidgetResolver` semantics). */
data class HeroCountdownPresentation(
    val labelKind: HeroCountdownLabelKind,
    val targetDate: Instant,
    val progressStartDate: Instant,
) {
    fun remainingSeconds(at: Instant): Int =
        maxOf(0, floor(Duration.between(at, targetDate).toMillis() / 1000.0).toInt())

    /** Elapsed fraction of `[progressStartDate, targetDate]` — matches accessory widget ring. */
    fun progress01(at: Instant): Double {
        val spanMillis = Duration.between(progressStartDate, targetDate).toMillis().toDouble()
        if (spanMillis <= 0) return 0.0
        val elapsedMillis = Duration.between(progressStartDate, at).toMillis().toDouble()
        val t = elapsedMillis / spanMillis
        return minOf(1.0, maxOf(0.0, t))
    }
}

sealed class PrayerEngineError : Exception() {
    data class NoIqamahRange(val day: Int) : PrayerEngineError() {
        override val message: String = "No iqamah range found for day $day"
    }

    data object MissingMonthly : PrayerEngineError() {
        private fun readResolve(): Any = MissingMonthly
        override val message: String = "missingMonthly"
    }

    data object MissingDayRow : PrayerEngineError() {
        private fun readResolve(): Any = MissingDayRow
        override val message: String = "missingDayRow"
    }

    data object MissingRamadanRow : PrayerEngineError() {
        private fun readResolve(): Any = MissingRamadanRow
        override val message: String = "missingRamadanRow"
    }
}
